sql/rowexec: TestUncertaintyErrorIsReturned failed

sql/rowexec.TestUncertaintyErrorIsReturned [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=3297637&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=3297637&tab=artifacts#/) on master @ [70183840a80160a652af8be84efd66430e4bc96e](https://github.com/cockroachdb/cockroach/commits/70183840a80160a652af8be84efd66430e4bc96e):


```
=== RUN   TestUncertaintyErrorIsReturned
    test_log_scope.go:73: test logs captured to: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestUncertaintyErrorIsReturned118819445
    test_log_scope.go:74: use -show-logs to present logs inline
=== CONT  TestUncertaintyErrorIsReturned
    processors_test.go:847: -- test log scope end --
test logs left over in: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestUncertaintyErrorIsReturned118819445
--- FAIL: TestUncertaintyErrorIsReturned (0.81s)
=== RUN   TestUncertaintyErrorIsReturned/vectorize=true/SELECT_*_FROM_t_AS_t1_JOIN_t_AS_t2_ON_t1.x_=_t2.x
    processors_test.go:842: 
        	Error Trace:	processors_test.go:842
        	            				processors_test.go:843
        	Error:      	Should be true
        	Test:       	TestUncertaintyErrorIsReturned/vectorize=true/SELECT_*_FROM_t_AS_t1_JOIN_t_AS_t2_ON_t1.x_=_t2.x
        	Messages:   	unexpected error: ERROR: inbox communication error: rpc error: code = Canceled desc = context canceled (SQLSTATE 58C01) running query SELECT * FROM t AS t1 JOIN t AS t2 ON t1.x = t2.x
        --- FAIL: TestUncertaintyErrorIsReturned/vectorize=true/SELECT_*_FROM_t_AS_t1_JOIN_t_AS_t2_ON_t1.x_=_t2.x (0.01s)
=== RUN   TestUncertaintyErrorIsReturned/vectorize=true
    --- FAIL: TestUncertaintyErrorIsReturned/vectorize=true (0.06s)
```
<details><summary>Reproduce</summary>
<p>
To reproduce, try:

```bash
make stressrace TESTS=TestUncertaintyErrorIsReturned PKG=./pkg/sql/rowexec TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```
Parameters in this failure:

- GOFLAGS=-parallel=4
</p>
</details>
/cc @cockroachdb/sql-queries asubiotto
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestUncertaintyErrorIsReturned.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

--- a/pkg/sql/colexec/BUILD.bazel
+++ b/pkg/sql/colexec/BUILD.bazel
@@ -61,7 +61,6 @@ go_library(
         "//pkg/sql/colexec/execgen",  # keep
         "//pkg/sql/colexecerror",
         "//pkg/sql/colexecop",
-        "//pkg/sql/colflow/colrpc",
         "//pkg/sql/colmem",
         "//pkg/sql/execinfra",
         "//pkg/sql/execinfrapb",

--- a/pkg/sql/colexec/parallel_unordered_synchronizer.go
+++ b/pkg/sql/colexec/parallel_unordered_synchronizer.go
@@ -20,7 +20,6 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/sql/colexec/colexecargs"
 	"github.com/cockroachdb/cockroach/pkg/sql/colexecerror"
 	"github.com/cockroachdb/cockroach/pkg/sql/colexecop"
-	"github.com/cockroachdb/cockroach/pkg/sql/colflow/colrpc"
 	"github.com/cockroachdb/cockroach/pkg/sql/execinfra"
 	"github.com/cockroachdb/cockroach/pkg/sql/execinfrapb"
 	"github.com/cockroachdb/cockroach/pkg/util"
@@ -68,12 +67,12 @@ type ParallelUnorderedSynchronizer struct {
 
 	inputs []colexecargs.OpWithMetaInfo
 	// cancelLocalInput stores context cancellation functions for each of the
-	// inputs. For each input tree, we examine whether it is fully local and
-	// only derive the cancellation function if so (i.e. there are no
-	// colrpc.Inboxes in the input tree).
-	cancelLocalInput     []context.CancelFunc
-	UsesFakeSpanResolver bool
-	tracingSpans         []*tracing.Span
+	// inputs. The functions are populated only if LocalPlan is true.
+	cancelLocalInput []context.CancelFunc
+	// LocalPlan indicates whether this synchronizer is a part of the fully
+	// local plan.
+	LocalPlan    bool
+	tracingSpans []*tracing.Span
 	// readNextBatch is a slice of channels, where each channel corresponds to the
 	// input at the same index in inputs. It is used as a barrier for input
 	// goroutines to wait on until the Next goroutine signals that it is safe to
@@ -170,17 +169,16 @@ func (s *ParallelUnorderedSynchronizer) Init(ctx context.Context) {
 	for i, input := range s.inputs {
 		var inputCtx context.Context
 		inputCtx, s.tracingSpans[i] = execinfra.ProcessorSpan(s.Ctx, fmt.Sprintf("parallel unordered sync input %d", i))
-		// If there are no colrpc.Inboxes in this input tree, the synchronizer
-		// can cancel the current work eagerly when transitioning into draining.
-		// If there is at least one inbox, then it cannot do so because
-		// canceling the context would break the gRPC stream and make it
-		// impossible to fetch the remote metadata.
-		//
-		// We also carve out an exception for fakedist* logic test configs since
-		// although the input tree might not contain an inbox, the tree is
-		// treated as remote, and we cannot cancel such tree eagerly.
-		localInput := !containsInbox(input) && !s.UsesFakeSpanResolver
-		if localInput {
+		if s.LocalPlan {
+			// If there plan is local, there are no colrpc.Inboxes in this input
+			// tree, and the synchronizer can cancel the current work eagerly
+			// when transitioning into draining.
+			//
+			// If there plan is distributed, there might be an inbox in the
+			// input tree, and the synchronizer cannot cancel the work eagerly
+			// because canceling the context would break the gRPC stream and
+			// make it impossible to fetch the remote metadata. Furthermore, it
+			// will result in the remote flow cancellation.
 			inputCtx, s.cancelLocalInput[i] = context.WithCancel(inputCtx)
 		}
 		input.Root.Init(inputCtx)
@@ -190,27 +188,6 @@ func (s *ParallelUnorderedSynchronizer) Init(ctx context.Context) {
 			}
 		}(input.Root, i)
 	}
-}
-
-// containsInbox returns true if there is at least one colrpc.Inbox in the
-// operator tree rooted in op.
-func containsInbox(op colexecargs.OpWithMetaInfo) bool {
-	// All inboxes must have been added to the MetadataSources, so rather than
-	// walking over the whole operator tree, we can just iterate over the
-	// MetadataSources.
-	for _, ms := range op.MetadataSources {
-		if util.CrdbTestBuild {
-			// If we have an InvariantsChecker around the MetadataSource, we
-			// have to unwrap it.
-			if i, ok := ms.(*InvariantsChecker); ok {
-				ms = i.Input.(colexecop.MetadataSource)
-			}
-		}
-		if _, ok := ms.(*colrpc.Inbox); ok {
-			return true
-		}
-	}
-	return false
 }
 
 func (s *ParallelUnorderedSynchronizer) getState() parallelUnorderedSynchronizerState {

--- a/pkg/sql/colexec/parallel_unordered_synchronizer_test.go
+++ b/pkg/sql/colexec/parallel_unordered_synchronizer_test.go
@@ -127,6 +127,7 @@ func TestParallelUnorderedSynchronizer(t *testing.T) {
 
 	var wg sync.WaitGroup
 	s := NewParallelUnorderedSynchronizer(inputs, &wg)
+	s.LocalPlan = true
 	s.Init(ctx)
 
 	t.Run(fmt.Sprintf("numInputs=%d/numBatches=%d/terminationScenario=%d", numInputs, numBatches, terminationScenario), func(t *testing.T) {

--- a/pkg/sql/colflow/vectorized_flow.go
+++ b/pkg/sql/colflow/vectorized_flow.go
@@ -954,7 +954,7 @@ func (s *vectorizedFlowCreator) setupInput(
 			// must use the serial unordered sync above in order to remove any
 			// concurrency.
 			sync := colexec.NewParallelUnorderedSynchronizer(inputStreamOps, s.waitGroup)
-			sync.UsesFakeSpanResolver = flowCtx.TestingKnobs().UsesFakeSpanResolver
+			sync.LocalPlan = flowCtx.Local
 			opWithMetaInfo = colexecargs.OpWithMetaInfo{
 				Root:            sync,
 				MetadataSources: colexecop.MetadataSources{sync},

--- a/pkg/sql/distsql_physical_planner.go
+++ b/pkg/sql/distsql_physical_planner.go
@@ -195,7 +195,6 @@ func (dsp *DistSQLPlanner) GatewayID() roachpb.NodeID {
 // responsibility to make sure the DistSQLPlanner is not in use.
 func (dsp *DistSQLPlanner) SetSpanResolver(spanResolver physicalplan.SpanResolver) {
 	dsp.spanResolver = spanResolver
-	dsp.distSQLSrv.TestingKnobs.UsesFakeSpanResolver = physicalplan.IsFakeSpanResolver(spanResolver)
 }
 
 // distSQLExprCheckVisitor is a tree.Visitor that checks if expressions

--- a/pkg/sql/execinfra/server_config.go
+++ b/pkg/sql/execinfra/server_config.go
@@ -234,10 +234,6 @@ type TestingKnobs struct {
 
 	// BackupRestoreTestingKnobs are backup and restore specific testing knobs.
 	BackupRestoreTestingKnobs base.ModuleTestingKnobs
-
-	// UsesFakeSpanResolver indicates whether the DistSQLPlanner uses a fake
-	// span resolver.
-	UsesFakeSpanResolver bool
 }
 
 // MetadataTestLevel represents the types of queries where metadata test

--- a/pkg/sql/physicalplan/fake_span_resolver.go
+++ b/pkg/sql/physicalplan/fake_span_resolver.go
@@ -40,12 +40,6 @@ func NewFakeSpanResolver(nodes []*roachpb.NodeDescriptor) SpanResolver {
 	}
 }
 
-// IsFakeSpanResolver returns whether s is a fakeSpanResolver.
-func IsFakeSpanResolver(s SpanResolver) bool {
-	_, ok := s.(*fakeSpanResolver)
-	return ok
-}
-
 // fakeRange indicates that a range between startKey and endKey is owned by a
 // certain node.
 type fakeRange struct {

